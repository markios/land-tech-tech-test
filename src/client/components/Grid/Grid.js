import React from 'react';
import GridCss from './Grid.module.css';

const Grid = ({ columns, rows, children, classNames }) => {
  const styles = [GridCss.Grid, classNames].join(' ');
  return (
    <div className={styles}>
      {children}
    </div>
  )
};

export default Grid;
