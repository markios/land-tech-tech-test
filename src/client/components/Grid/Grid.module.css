
.Grid {
  background-size: calc(100% / 100) calc(100% / 100);
  background-image: radial-gradient(circle, var(--main-colour) 1px, rgba(0, 0, 0, 0) 1px);
  position: relative;
  margin: 0 auto;
  display: grid;
  grid-template-columns: repeat(100, 1fr);
  grid-template-rows: repeat(100, 1fr);
  grid-gap: 0;
}

.Grid__item {
  display: flex;
  align-items: center;
  justify-content: center;
}