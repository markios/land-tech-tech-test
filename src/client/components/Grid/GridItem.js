import React from 'react';
import GridCss from './Grid.module.css';

const GridItem = ({x, y, children}) => { 
  return (<div className={GridCss.Grid__item} style={({gridArea: `${x} / ${y}`})}>{children}</div>);
}

export default GridItem;