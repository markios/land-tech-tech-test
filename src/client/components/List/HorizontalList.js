import React from 'react';
import ListCss from './List.module.css';

const HorizontalList = ({children, classNames = ''}) => {
  const styles = [ListCss.List, ListCss['List--horizontal'], classNames].join(' ');
  return (
    <ul className={styles}>
      {React.Children.map(children, (child, i) => (<li>{child}</li>))}
    </ul>
  )
}

export default HorizontalList;