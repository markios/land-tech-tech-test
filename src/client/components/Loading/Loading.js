import React from 'react';
import LoadingCss from './Loading.module.css';
import FontsCss from '../../common/style/fonts.module.css';
import ColoursCss from '../../common/style/colours.module.css';

const Loading = () =>  {
  const styles = [LoadingCss.Loading, FontsCss.mainFont, ColoursCss.colourSecondary, FontsCss.fontSizeL].join(' ');
  return (<div className={styles}>Loading</div>);
}

export default Loading;