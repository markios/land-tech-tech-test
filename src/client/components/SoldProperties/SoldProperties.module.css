.SoldProperties {
  min-height: 100%;
  margin: 50px auto 100px;
  width: 100%;
  max-width: 1200px;
}

.SoldProperties__Grid {
  position: relative;
}

.SoldProperties__List {
  display: flex;
  flex-direction: row;
  margin-bottom: 20px;
}

.SoldProperties__List__Menu {
  margin: 0 0 0 auto;
}

.SoldProperties__List__Menu__Item {
  display: flex;
  align-items: center;
}

.SoldProperties__List__Menu__Item span {
  margin: 0 5px 0;
}