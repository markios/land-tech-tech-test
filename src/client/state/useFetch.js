// hooks.js
import { useState, useEffect } from "react";

function useFetch(url) {
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [failed, setFailed] = useState(false);
  
  useEffect(() => {
    async function fetchUrl() {
      try {
        const response = await fetch(url);
        const json = await response.json();
        setData(json);
      } catch (error) {
        setFailed(true);
      } finally {
        setLoading(false);
      }
    }
    fetchUrl();
  }, [url]);

  return [data, loading, failed];
}
export { useFetch };