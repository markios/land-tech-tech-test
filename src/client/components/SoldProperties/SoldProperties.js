import React from 'react';
import { useFetch } from '../../state/useFetch';
import {
  mapPropertiesForGroup,
  formatToLocalCurrency,
} from '../../state/selectors';
import constants from '../../state/constants';
import Grid from '../Grid/Grid';
import Loading from '../Loading/Loading';
import IconHouse from '../Icons/IconsHouse';
import HorizontalList from '../List/HorizontalList';
import GridItem from '../Grid/GridItem';
import Tooltip from '../Tooltip/Tooltip';
import Colours from '../../common/style/colours.module.css';
import SoldCss from './SoldProperties.module.css';
import FontsCss from '../../common/style/fonts.module.css';
import IconStyles from '../Icons/Icons.module.css';

const propertiesCharColours = {
  [constants.ZERO_TO_FIVE]: Colours.fillGroupOne,
  [constants.FIVE_TO_TWENTY_FIVE]: Colours.fillGroupTwo,
  [constants.TWENTY_FIVE_TO_SEVENTY_FIVE]: Colours.fillGroupThree,
  [constants.SEVENTY_FIVE_TO_NINETY_FIVE]: Colours.fillGroupFour,
  [constants.NINETY_FIVE_TO_ONE_HUNDRED]: Colours.fillGroupFive,
};

const propertiesNameSet = {
  [constants.ZERO_TO_FIVE]: '0 - 5%',
  [constants.FIVE_TO_TWENTY_FIVE]: '5% - 25%',
  [constants.TWENTY_FIVE_TO_SEVENTY_FIVE]: '25% - 75%',
  [constants.SEVENTY_FIVE_TO_NINETY_FIVE]: '75% - 95%',
  [constants.NINETY_FIVE_TO_ONE_HUNDRED]: '95% - 100%',
};

const SoldProperties = () => {
  const [data, loading, failed] = useFetch('/api/property/sold');
  const hasLoaded = !loading && !failed;
  const listStyles = [FontsCss.mainFont, FontsCss.fontSizeS, SoldCss.SoldProperties__List].join(' ');
  const errorStyles = [FontsCss.mainFont, FontsCss.fontSizeM, Colours.coloursError].join(' ');
  return (
    <section className={SoldCss.SoldProperties}>
      <div className={listStyles}>
        {hasLoaded &&
          <div role="contentinfo">
            <strong>Total Properties</strong>: {data.metadata.totalCount}
          </div>
        }
        <HorizontalList classNames={SoldCss.SoldProperties__List__Menu}>
          {Object.entries(propertiesCharColours).map(([key, colour]) => {
            return (
              <div key={`house_${key}`} className={SoldCss.SoldProperties__List__Menu__Item}>
                <IconHouse classNames={`${colour} ${IconStyles['Icons__item--lg']}`} />
                <span>{propertiesNameSet[key]}</span>
              </div>
            )
          })}
        </HorizontalList>
      </div>
      {loading && <Loading />}
      {failed && <p className={errorStyles}>Failed to fetch data</p>}
      {hasLoaded &&
        <Grid classNames={SoldCss.SoldProperties__Grid}>
          {mapPropertiesForGroup(data).map(({x, y, price, group}) => (
            <GridItem key={`${x}_${y}`} x={x} y={y}>
              <Tooltip content={formatToLocalCurrency(price)}>
                <IconHouse classNames={propertiesCharColours[group]} />
              </Tooltip>
            </GridItem>
          ))}
        </Grid>
      }
    </section>
  )
};

export default SoldProperties;
