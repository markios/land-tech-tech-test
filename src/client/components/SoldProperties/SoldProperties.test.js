import React from 'react';
import { shallow } from 'enzyme';
import { useFetch } from '../../state/useFetch';
import Loading from '../Loading/Loading';
import GridItem from '../Grid/GridItem';
import Grid from '../Grid/Grid';
import HorizontalList from '../List/HorizontalList';
import SoldProperties from './SoldProperties';

jest.mock('../../state/useFetch', () => ({
  useFetch: jest.fn(),
}));

describe('Tooltip', () => {
  let data;
  let loading;
  let failed;
  beforeEach(() => {
    data = [];
    loading = true;
    failed = false;
    useFetch.mockImplementation(() => {
      return [data, loading, failed];
    });
  });
  
  it('renders without crashing', () => {
    shallow(<SoldProperties />);
  });

  it('renders menu guide', () => {
    const wrapper = shallow(<SoldProperties />);

    const list = wrapper.find(HorizontalList);
    expect(list).toHaveLength(1);
    expect(list.html()).toMatchSnapshot();
  });

  it('shows loading while fetching', () => {
    const wrapper = shallow(<SoldProperties />);

    expect(wrapper.contains(<Loading />)).toEqual(true);
  });

  it('shows error if useFetch fails', () => {
    loading = false;
    failed = true;
    const wrapper = shallow(<SoldProperties />);
    
    expect(wrapper.html().includes("Failed to fetch data")).toEqual(true);
  });

  it('correctly renders total items', () => {
    loading = false;
    const records = [{
      x: 11,
      y: 22,
      price: 100000
    }];

    data = { 
      records,
      metadata: {
        totalCount: records.length
      } 
    };
    const wrapper = shallow(<SoldProperties />);

    expect(wrapper.find('[role="contentinfo"]').html()).toMatchSnapshot();
  });

  it('correctly renders item in grid', () => {
    loading = false;
    const records = [{
      x: 11,
      y: 22,
      price: 100000
    }, {
      x: 1,
      y: 2,
      price: 99999
    }];

    data = { 
      records,
      metadata: {
        totalCount: records.length
      } 
    };
    const wrapper = shallow(<SoldProperties />);
    
    expect(wrapper.find(GridItem).length).toEqual(records.length);
    expect(wrapper.find(Grid).html()).toMatchSnapshot();
  });
});