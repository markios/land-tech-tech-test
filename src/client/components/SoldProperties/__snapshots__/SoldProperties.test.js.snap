// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip correctly renders item in grid 1`] = `"<div class=\\"Grid SoldProperties__Grid\\"><div class=\\"Grid__item\\" style=\\"grid-area:11 / 22\\"><div data-content=\\"£100,000\\" class=\\"Tooltip fontSizeXS\\"><svg class=\\"Icons__item fillGroupFive\\"><use xlink:href=\\"#icon-house\\"></use></svg></div></div><div class=\\"Grid__item\\" style=\\"grid-area:1 / 2\\"><div data-content=\\"£99,999\\" class=\\"Tooltip fontSizeXS\\"><svg class=\\"Icons__item fillGroupFive\\"><use xlink:href=\\"#icon-house\\"></use></svg></div></div></div>"`;

exports[`Tooltip correctly renders total items 1`] = `"<div role=\\"contentinfo\\"><strong>Total Properties</strong>: 1</div>"`;

exports[`Tooltip renders menu guide 1`] = `"<ul class=\\"List List--horizontal SoldProperties__List__Menu\\"><li><div class=\\"SoldProperties__List__Menu__Item\\"><svg class=\\"Icons__item fillGroupOne Icons__item--lg\\"><use xlink:href=\\"#icon-house\\"></use></svg><span>0 - 5%</span></div></li><li><div class=\\"SoldProperties__List__Menu__Item\\"><svg class=\\"Icons__item fillGroupTwo Icons__item--lg\\"><use xlink:href=\\"#icon-house\\"></use></svg><span>5% - 25%</span></div></li><li><div class=\\"SoldProperties__List__Menu__Item\\"><svg class=\\"Icons__item fillGroupThree Icons__item--lg\\"><use xlink:href=\\"#icon-house\\"></use></svg><span>25% - 75%</span></div></li><li><div class=\\"SoldProperties__List__Menu__Item\\"><svg class=\\"Icons__item fillGroupFour Icons__item--lg\\"><use xlink:href=\\"#icon-house\\"></use></svg><span>75% - 95%</span></div></li><li><div class=\\"SoldProperties__List__Menu__Item\\"><svg class=\\"Icons__item fillGroupFive Icons__item--lg\\"><use xlink:href=\\"#icon-house\\"></use></svg><span>95% - 100%</span></div></li></ul>"`;
